profiles/consensus/scripts/nbconverted/build-consensus-signatures.py

```diff
@@ -0,0 +1,226 @@
+            df.groupby(replicate_cols)
+            .apply(lambda x: modz_base(x.loc[:, cp_features]))
+            .reset_index()
+        )
+    elif operation == "median":
+        consensus_df = aggregate(
+            df, operation="median", features="infer", strata=replicate_cols
+        )
+    return consensus_df
+
+
+# In[4]:
+
+
+# Set constants
+file_base = "_normalized_dmso.csv.gz"
+operations = ["median", "modz"]
+batch = "2016_04_01_a549_48hr_batch1"
+primary_dose_mapping = [0.04, 0.12, 0.37, 1.11, 3.33, 10, 20]
+
+# Set file paths
+profile_dir = pathlib.Path("..", batch)
+plate_dirs = [x for x in profile_dir.iterdir() if x.name != ".DS_Store"]
+plates = [x.name for x in plate_dirs]
+
+
+# In[5]:
+
+
+# The output directory is also the batch name
+pathlib.Path(batch).mkdir(exist_ok=True)
+
+
+# ## Load Profile Data Per Plate and Concatenate
+
+# In[6]:
+
+
+# Load Data
+all_profiles_df = []
+for plate_dir in plate_dirs:
+    plate = plate_dir.name
+    plate_file = plate_dir / f"{plate}{file_base}"
+    plate_df = pd.read_csv(plate_file)
+    all_profiles_df.append(plate_df)
+
+# Concatenate profiles
+all_profiles_df = pd.concat(all_profiles_df, axis="rows")
+
+# Determine every CellProfiler feature measured
+cp_features = infer_cp_features(all_profiles_df)
+
+print(all_profiles_df.shape)
+all_profiles_df.head()
+
+
+# ## Recode Dose
+
+# In[7]:
+
+
+# Recode dose
+all_profiles_df = all_profiles_df.assign(
+    Metadata_dose_recode=(
+        all_profiles_df.Metadata_mmoles_per_liter.apply(
+            lambda x: recode_dose(x, primary_dose_mapping, return_level=True)
+        )
+    )
+)
+
+# Make sure DMSO profiles recieve a zero dose level
+all_profiles_df.loc[
+    all_profiles_df.Metadata_broad_sample == "DMSO", "Metadata_dose_recode"
+] = 0
+
+
+# In[8]:
+
+
+all_profiles_df.Metadata_dose_recode.value_counts()
+
+
+# In[9]:
+
+
+all_profiles_df.plot(
+    x="Metadata_mmoles_per_liter", y="Metadata_dose_recode", kind="scatter"
+)
+
+
+# ## Create Consensus Profiles
+# 
+# ### a) Generate consensus profiles for DMSO by well
+
+# In[10]:
+
+
+dmso_replicate_cols = [
+    "Metadata_broad_sample",
+    "Metadata_pert_well",
+    "Metadata_mmoles_per_liter",
+    "Metadata_dose_recode",
+]
+
+# Isolate DMSO profiles
+dmso_profile_df = all_profiles_df.query("Metadata_broad_sample == 'DMSO'").reset_index(
+    drop=True
+)
+
+# How many DMSO profiles per well?
+dmso_profile_df.Metadata_pert_well.value_counts()
+
+
+# In[11]:
+
+
+# Form dmso consensus profiles with median and modz
+dmso_profiles = {}
+for operation in operations:
+    dmso_profiles[operation] = consensus_apply(
+        dmso_profile_df,
+        operation=operation,
+        cp_features=cp_features,
+        replicate_cols=dmso_replicate_cols,
+    )
+
+
+# ### b) Generate consensus profiles per compound-dose pair
+
+# In[12]:
+
+
+compound_replicate_cols = [
+    "Metadata_broad_sample",
+    "Metadata_mmoles_per_liter",
+    "Metadata_dose_recode",
```

Comment:
We should also include `Metadata_Plate_Map_Name` here. This makes is easier to detect batch effects with the small handful (n=2 I think) of compounds present on multiple platemaps.

Same thing with `dmso_replicate_cols`. You will get 24 wells x 28 plate maps = 672 DMSO profiles, which is what we want.

Reply:
this is great!! Thanks for catching